File: src/main/java/preponderous/viron/services/GridService.java
// Copyright (c) 2024 Preponderous Software
// MIT License

package preponderous.viron.services;

import java.util.Arrays;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import preponderous.viron.config.ServiceConfig;
import preponderous.viron.models.Grid;

@Service
public class GridService {
    private final RestTemplateBuilder restTemplateBuilder;
    private final ServiceConfig serviceConfig;

    private final String baseUrl;
    
    @Autowired
    public GridService(RestTemplateBuilder restTemplateBuilder, ServiceConfig serviceConfig) {
        this.restTemplateBuilder = restTemplateBuilder;
        this.serviceConfig = serviceConfig;

        this.baseUrl = this.serviceConfig.getVironHost() + ":" + serviceConfig.getVironPort() + "/grid";
    }

    public List<Grid> getAllGrids() {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Grid[]> response = restTemplate.exchange(baseUrl + "/get-all-grids", HttpMethod.GET, null, Grid[].class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting grids");
        }
        return Arrays.asList(response.getBody());
    }

    public Grid getGridById(int id) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Grid> response = restTemplate.exchange(baseUrl + "/get-grid-by-id/" + id, HttpMethod.GET, null, Grid.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting grid by id");
        }
        return response.getBody();
    }

    public List<Grid> getGridsInEnvironment(int environmentId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Grid[]> response = restTemplate.exchange(baseUrl + "/get-grids-in-environment/" + environmentId, HttpMethod.GET, null, Grid[].class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting grids in environment");
        }
        return Arrays.asList(response.getBody());
    }

    public Grid getGridOfEntity(int entityId) {
        RestTemplate restTemplate = restTemplateBuilder.build();
        ResponseEntity<Grid> response = restTemplate.exchange(baseUrl + "/get-grid-of-entity/" + entityId, HttpMethod.GET, null, Grid.class);
        if (response.getStatusCode().isError()) {
            throw new RuntimeException("Error getting grid of entity");
        }
        return response.getBody();
    }

    // testing
    public static void main(String[] args) {
        ServiceConfig serviceConfig = new ServiceConfig();
        serviceConfig.setVironHost("http://localhost");
        serviceConfig.setVironPort(8080);
        GridService gridService = new GridService(new RestTemplateBuilder(), serviceConfig);
        List<Grid> grids = gridService.getGridsInEnvironment(1);
        for (Grid grid : grids) {
            System.out.println(grid);
        }
    }
}

File: src/main/python/preponderous/viron/services/gridService.py
# Copyright (c) 2024 Preponderous Software
# MIT License

import logging
from typing import List, Optional
import requests
from requests.exceptions import HTTPError
from src.main.python.preponderous.viron.models.grid import Grid

logger = logging.getLogger(__name__)

class GridService:
    def __init__(self, base_url: str):
        self.base_url = f"{base_url}/api/v1/grids"

    def get_all_grids(self) -> List[Grid]:
        response = requests.get(self.base_url)
        response.raise_for_status()
        return [Grid(**grid) for grid in response.json()]

    def get_grid_by_id(self, grid_id: int) -> Optional[Grid]:
        response = requests.get(f"{self.base_url}/{grid_id}")
        response.raise_for_status()
        return Grid(**response.json())

    def get_grids_in_environment(self, environment_id: int) -> List[Grid]:
        response = requests.get(f"{self.base_url}/environment/{environment_id}")
        response.raise_for_status()
        return [Grid(**grid) for grid in response.json()]

    def get_grid_of_entity(self, entity_id: int) -> Optional[Grid]:
        response = requests.get(f"{self.base_url}/entity/{entity_id}")
        response.raise_for_status()
        return Grid(**response.json())